File: 200 Number of Islands.c
"""
Created on Mon Nov. 09 15:39 2020

@author: Chiwen Chen
"""

int numIslands(char** grid, int gridSize, int* gridColSize){
    int i,j;
    int num=0;
    for(i=0; i < gridSize;i++){
        for(j=0; j < gridColSize[i]; j++){
            if(grid[i][j]=='1'){
                visit(grid,i,j,gridSize,gridColSize);
                num=num+1;
            }
            
        }
        
    }
    return num;

}

void visit(char **grid, int i, int j, int gridSize, int *gridColSize)
{
    if (i < 0 || j < 0 || i >= gridSize)
        return;
    
    if (j >= gridColSize[i])
        return;
    
    if (grid[i][j] == '0')
        return;
    
    grid[i][j] = '0';
    
    visit(grid, i-1, j, gridSize, gridColSize);//<
    visit(grid, i+1, j, gridSize, gridColSize);//>
    visit(grid, i, j-1, gridSize, gridColSize);//up
    visit(grid, i, j+1, gridSize, gridColSize);//down
}
//Runtime: 12 ms, faster than 77.52% of C online submissions for Number of Islands.
//Memory Usage: 7.8 MB, less than 5.05% of C online submissions for Number of Islands.

File: 200 Number of Islands.py
"""
Created on Mon Nov. 09 16:28 2020

@author: Chiwen Chen
"""

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if len(grid) == 0: return 0
    
        row , col  = len(grid), len(grid[0])
        ans = 0
        for i in range(row):
            for j in range(col):
                if grid[i][j] == '1':
                    q = [(i, j)]
                    while q:
                        x, y = q.pop()
                        """point out"""
                        grid[x][y] = '2'
                        """"search near"""
                        for sx, sy in (0, 1), (0, -1), (1, 0), (-1, 0):
                            new_x, new_y = x+sx, y+sy
                            if 0 <= new_x < row and 0 <= new_y < col and grid[new_x][new_y] == '1':
                                q.append((new_x, new_y))
                    ans += 1            
        return ans
    """
    Runtime: 128 ms, faster than 92.31% of Python3 online submissions for Number of Islands.
    Memory Usage: 15 MB, less than 5.62% of Python3 online submissions for Number of Islands.
    """